Implements service modules registry - Experimental

dmarteau opened the pull request:
Implementation for https://github.com/qgis/QGIS-Enhancement-Proposals/issues/74

~~This is an experimental work in progress.~~
 
~~The interfaces need refinements and a native (C++) exemple implementation of module is missing.~~
I have  issue with dynamic modules on mac:  after make 'install' the runtime path/install name of qgis_server lib is set to the build location and not the install location, so module that link dynamically are also linked with libs in the build location not install location. 
Help is welcome for setting the  build process of dynamic modules. 

What is done so far:

- Handling of services as dynamic modules (native and python)
- Découpling of the fcgi supports from the server codebase.  
- WMS supports is moved to new service provider system.

dmarteau committed:
Server refactoring: implements service modules registry

    Implementation for https://github.com/qgis/QGIS-Enhancement-Proposals/issues/74

m-kuhn commented on Testing/Temporary/CTestCostData.txt:
```diff
@@ -0,0 +1 @@
+---
```
Should this file really be in here or rather be added to .gitignore?

dmarteau replied:
It looks like an artefact created from test run.

dmarteau replied:
It seems that the whole Testing directory should be added to .gitignore.

m-kuhn replied:
https://github.com/qgis/QGIS/pull/3868 does exactly this

m-kuhn replied:
Can you remove this file? The .gitignore part is already merged.

rldhont commented:
@pblottiere this is the PR for QGIS Server modularity

nyalldawson commented on python/server/qgsserverrequest.sip:
```diff
@@ -0,0 +1,76 @@
+       /**
+        * @return  the request url
+        */
+       virtual const QUrl& url() const; 
```
Qurl is implicitly shared, so no need to return a reference here.

nyalldawson commented on python/server/qgsserverrequest.sip:
```diff
@@ -0,0 +1,76 @@
+/**
+ * \ingroup server
+ * QgsServerRequest
+ * Class defining request intreface passed to services QgsService::executeRequest() method
```
Interface

nyalldawson commented on python/server/qgsserverrequest.sip:
```diff
@@ -0,0 +1,76 @@
+/***************************************************************************
+                          qgsserverrequest.h
+
+  Define ruquest class for getting request contents
```
Request

nyalldawson commented on python/server/qgsserverrequest.sip:
```diff
@@ -0,0 +1,76 @@
+       virtual const QUrl& url() const; 
+
+       /**
+        * @return the rquest method
```
Request

nyalldawson commented on python/server/qgsserverrequest.sip:
```diff
@@ -0,0 +1,76 @@
+        * The default implementation retfurn nullptr
+        * @return a QByteArray pointer or nullptr
+        */
+       virtual const QByteArray* data() const;
```
Qbytearray is also implicit shared, so just return a value here.

dmarteau replied:
@nyalldawson  I'm confused: the QByteArray hold post/put  data and may not exists at all.

dmarteau replied:
@m-kuhn  So the default implementation should be 'return QByteArray()'  which will be a null QByteArray ?

m-kuhn replied:
Sounds reasonable to me

dmarteau replied:
Ok

nyalldawson commented on python/server/qgsserverresponse.sip:
```diff
@@ -0,0 +1,77 @@
+ * QgsServerResponse
+ * Class defining response interface passed to services QgsService::executeRequest() method
+ *
+ *  Note:
```
These notes don't relate to api usage, so should be pulled out of the doxygen tags and just put in normal c++ comments

nyalldawson commented on python/server/qgsserverresponse.sip:
```diff
@@ -0,0 +1,77 @@
+
+       /** Set header entry 
+        *  Add header entry to the response 
+        *  Note that it is usually an error to set hedaer after writng data
```
Header

nyalldawson commented on python/server/qgsserverresponse.sip:
```diff
@@ -0,0 +1,77 @@
+         */
+        virtual void sendError( int code,  const QString& message ) = 0;
+
+        /**
```
These docs could be more verbose. It's quite obvious that this method writes a string from its signature, but the docs should at least say where the string is written to.

nyalldawson commented on python/server/qgsserverresponse.sip:
```diff
@@ -0,0 +1,77 @@
+        virtual void write(const QString& data );
+
+       /**
+        * Write chunk af data
```
Af -> of. Maybe run this whole pr through a spell checker ;)

nyalldawson commented on src/server/CMakeLists.txt:
```diff
@@ -50,6 +50,16 @@ SET ( qgis_mapserv_SRCS
   qgssldconfigparser.cpp
   qgsconfigparserutils.cpp
   qgsserver.cpp
+#XXX https://github.com/qgis/QGIS-Enhancement-Proposals/issues/74
```
No need for this extra comment here

nyalldawson commented on src/server/qgsserverresponse.h:
```diff
@@ -0,0 +1,86 @@
+        * They are convenience method that will write directly to the 
+        * underlying I/O device
+        */
+        virtual qint64 write( const QByteArray &byteArray );
```
Why not just int return values here? What's the need for qint64?

dmarteau replied:
These methods are a direct mapping of QIODevice::write they have the same return value.

nyalldawson commented on src/server/qgsserverresponse.h:
```diff
@@ -0,0 +1,86 @@
+        virtual qint64 write( const QByteArray &byteArray );
+
+        // Not exposed in python
+        virtual qint64 write( const char* data, qint64 maxsize);
```
Should be

    //! @note not available in pything bindings

Proper documentation about what the c++ method does should also be added

nyalldawson commented on src/server/qgsservice.h:
```diff
@@ -0,0 +1,64 @@
+    /**
+     * Return true if the given method is supported for that
+     * service.
+     * @param method QGSMethodType the 
```
What's a QGSMethodType? Is this an old comment?

dmarteau replied:
Yes its an old comment. fixed.

nyalldawson commented on src/server/qgsservice.h:
```diff
@@ -0,0 +1,64 @@
+     * Return true if the given method is supported for that
+     * service.
+     * @param method QGSMethodType the 
+     * @return QString containing the configuration file path
```
This line should be removed or updated

nyalldawson commented on src/server/qgsserviceloader.h:
```diff
@@ -0,0 +1,57 @@
+    QgsServiceLoader();
+
+    //! Destructor
+    virtual ~QgsServiceLoader() = 0;
```
Is this class intended to be inherited? If so, the pure virtual destructor could be a problem for python subclasses.

dmarteau replied:
This class is internal  is not intended to be exposed in python.  
The constructor should not be needed but actually I found that  if a constructor was not defined then linker complains about missing symbol when using the python binding library.

nyalldawson replied:
In that case the class dox will need a "\note not available in Python bindings" added.

nyalldawson commented on src/server/qgsserviceregistry.cpp:
```diff
@@ -0,0 +1,185 @@
+namespace {
+
+    // Build a key entry from name and version
+    QString make_service_key( const QString& name, const QString& version  )
```
Qgis naming convention is camel case.

nyalldawson commented on src/server/qgsserviceregistry.cpp:
```diff
@@ -0,0 +1,185 @@
+    QString key;
+
+    // Check that we have a service of that name
+    VersionTable::iterator v = mVersions.find(name);
```
This should probably work with constFind/constEnd . Will be a bit faster.

nyalldawson commented on src/server/qgsserviceregistry.cpp:
```diff
@@ -0,0 +1,185 @@
+{
+    // Test if service is already registered
+    QString key = make_service_key( name, version );
+    if( mServices.find(key) != mServices.end() )
```
Same here - use constFind/constBegin/constEnd wherever possible

nyalldawson commented on src/server/qgsserviceregistry.h:
```diff
@@ -0,0 +1,105 @@
+     * @param service a QgsServerResponse to be registered
+     * @param version the version string for the service  (required)
+     */
+    void registerService( const QString& name,  QgsService* service, const QString& version );
```
Name/version/service would be a more natural order

nyalldawson reviewed (commented):
@dmarteau I'm not involved in server so can't comment on the design, so this feedback is just general QGIS conventions/qt usage advice.

Great work here!

wonder-sk commented on python/server/qgsserverrequest.sip:
```diff
@@ -0,0 +1,76 @@
+ *  Note about design: this intreface must be passed along to python and thus signatures methods must be 
+ *  compatible with pyQGS/pyQT api and rules.
+ */
+class QgsServerRequest
```
How about headers sent with the request... it might be useful to have them available...?

wonder-sk commented on python/server/qgsserverresponse.sip:
```diff
@@ -0,0 +1,77 @@
+        /**
+         * Write string 
+         */
+        virtual void write(const QString& data );
```
Not sure if this method should be in the interface... IMO we should be dealing here only with byte arrays. It is unclear what encoding would be used for conversion from string to byte array.

dmarteau replied:
This method was intended to be a facility method for direcly sending string (UTF-8) to the underlying device. The default implementation creates a QTextStream around the QIOdevice and sent the QString. It is very convenient in python where you always need to instanciate a QTextStream for sending text.

wonder-sk replied:
I understand the intent, but doing something like response.write(some_string.encode('utf-8')) is easier to read and understand than response.writeString(some_string) in my opinion...

dmarteau replied:
mmmh, I'm not sure I still want to deal with encode/decode in python 3 :-)

wonder-sk commented on python/server/qgsserverresponse.sip:
```diff
@@ -0,0 +1,77 @@
+        /**
+         * Return the underlying QIODevice
+         */
+        virtual QIODevice* io() = 0;
```
When would someone want to use the underlying I/O device directly? Worth documenting...

mhugo replied:
I agree.
What about a QgsServerResponse that inherits from QIODevice ?

dmarteau replied:
The QIODevice is the base class for I/O operation, it provides device-independent I/O features . It can be wrapped with all Stream classes (QDataStream, QTextStream...) so giving access to it enable services implementors to use the appropriate streaming object 
depending on the the type of the response. 
Furthemore the QgsServerResponse must support streaming of data and the QIODevice is appropriate for that.

dmarteau replied:
@mhugo Because a QgsResponse is not a I/O device and you could not use QBuffer, QFile, QSocket which are implementatian of QIODevice. Or if you delegate yould would have to map all QIODEvice methods each time you need to create a  QgsServerResponse class. Have a look to the python tests.

wonder-sk commented on python/server/qgsserverresponse.sip:
```diff
@@ -0,0 +1,77 @@
+        * They are convenience method that will write directly to the 
+        * underlying I/O device
+        */
+        virtual qint64 write(const QByteArray& byteArray );
```
Does the client need to check the return value and is it an error if the returned value is not equal to the length of the byte array?

dmarteau replied:
These methods are a direct mapping of QIODevice::write, so yes.
 I'm fixing the documentation.

wonder-sk commented on python/server/qgsserviceregistry.sip:
```diff
@@ -0,0 +1,83 @@
+ *
+ * This class provides methods for registering and retrieving 
+ * services. Note that the regstiry does not hord ownership of
+ * registered service but vill call the 'release' method on cleanup)
```
If I understand the comment correctly, this comment is in contradiction to what registerService() docs say about ownership of services...

dmarteau replied:
Yes, that comment is out of date. It is fixed.

DelazJ replied:
typo: will

wonder-sk commented on python/server/qgsserviceregistry.sip:
```diff
@@ -0,0 +1,83 @@
+     * @param version the required version (optional)
+     * @return QgsService
+     */
+    QgsService* getService( const QString& name, const QString& version = QString() );
```
How does the mechanism of versions work here? Isn't the versioning just something specific to OGC services that could be handled inside the service implementation rather than here?

dmarteau replied:
The version mechanism is not limited to version as defined for ogc service: we want to be able to use or test  services under development  or with different implementations. 
Actually if no version is provided the highest version win. But if you specify a version that it not registered the method fail (return null). 
Note that a 'module' may register several versions of a service and thus delegate to the same implementation.
There is room for discussion...

wonder-sk replied:
I would very much like to leave the versioning mechanism to individual services. So there would be one WMS service that handles all WMS requests. Often the changes between protocol versions are quite small, so it is useful to have one service for multiple versions.

rldhont replied:
Hi @wonder-sk , but sometimes there are big differences like in WFS 1 and 2, and some devs would like to test or implement different version in different way. But this was a hard discussion during code sprint.

dmarteau replied:
Has I said before, module can register multiples instances of the same service but with different version so they can share the same code.

wonder-sk commented on python/server/qgsserviceregistry.sip:
```diff
@@ -0,0 +1,83 @@
+     * If pythonModulePath is not specified the environnement variables QGIS_PYTHON_SERVICE_PATH
+     * is examined. 
+     */
+    void init( const QString& nativeModulepath, const QString& pythonModulePath = QString() );
```
Do we need init() and cleanUp() methods? Maybe their code could be in constructor/destructor of the class?

dmarteau replied:
IMHO there is no correlation between the lifetime of the QgsServiceRegistry and the lifetime of the services. Providing init() and cleanup() enable decoupling instanciation from functional behavior.

wonder-sk commented on python/server/qgsserviceregistry.sip:
```diff
@@ -0,0 +1,83 @@
+#include "qgsserviceregistry.h"
+#include "qgsservice.h"
+#include "qgsserverrequest.h"
+#include "qgsserverresponse.h"
```
It should not be necessary to #include also request/response headers here...

wonder-sk commented on python/server/qgsserviceregistry.sip:
```diff
@@ -0,0 +1,83 @@
+     * @param service a QgsServerResponse to be registered
+     * @param version the service version 
+     */
+    void registerService( const QString& name,  QgsService* service /Transfer/, const QString& version );
```
Would be nice to also have unregisterService() method - for example if a python plugin would be unloaded/reloaded...

dmarteau replied:
Yes unRegisterService is probably missing (at least for the sake of completeness). But in the context of a server do we want the ability to change the configuration at runtime ?

wonder-sk commented on python/server/qgsserviceregistry.sip:
```diff
@@ -0,0 +1,83 @@
+ * registered service but vill call the 'release' method on cleanup)
+ *
+ */
+class QgsServiceRegistry
```
It would be also useful to have a method to get a list of registered services...

wonder-sk commented on src/server/qgsserviceregistry.cpp:
```diff
@@ -0,0 +1,185 @@
+       QString     mName;
+       QgsService* mService;
+       QString     mVersion;    
+};
```
Maybe these bits of information like name/version could be already inside of QgsService class, so there is no need for service entry classes?

dmarteau replied:
I wanted to tave the QgsService intreface as small as possible but you are probably right.

dmarteau replied:
I have moved name version in QgsService and removed the entry classe.

wonder-sk commented on python/server/qgsserverrequest.sip:
```diff
@@ -0,0 +1,76 @@
+       QgsServerRequest( const QUrl& url, Method method );
+
+       //! destructor
+       virtual ~QgsServerRequest();
```
Why are the request methods virtual? Is it expected that others would make subclasses of it?

wonder-sk reviewed (commented):
Nice work - well done - looking for more :-)

Do I understand correctly that the request/response classes will have some implementations provided in the server (e.g. impl based on QNetworkRequest/QNetworkReply and impl working with just memory buffers) ?

dmarteau commented:
@wonder-sk  This is exactly that. Thanks.

dmarteau committed:
Server refactoring: Fix doc and comments

    Fix typo in doc and comments
    Update documentation
    Fix return of implicitely shared objects
    Fix order of parameters in QgsServiceRegistry::registerService

dmarteau committed:
Remove Service Entry class and move name/version infos in QgsService

dmarteau committed:
Add getHeader() method to QgsServerRequest

mhugo commented on python/server/qgsserverresponse.sip:
```diff
@@ -0,0 +1,80 @@
+         * @param code HHTP return code value
+         * @param message An informative error message
+         */
+        virtual void sendError( int code,  const QString& message ) = 0;
```
I am not sure to see the difference with setReturnCode

dmarteau replied:
@mhugo you can return legitimate response body with different return code (ex: 200 to 206 codes) so you  should have the possibility to set a return code wich does not necessary indicates an error. Also, you should be able to return a true error code without having to deal with the particular format of the error response which must be handled uniformly across all services.

DelazJ commented on python/server/qgsserverresponse.sip:
```diff
@@ -0,0 +1,80 @@
+
+       /** Set Header entry 
+        *  Add Header entry to the response 
+        *  Note that it is usually an error to set Header after writng data
```
writing

DelazJ commented on python/server/qgsserverresponse.sip:
```diff
@@ -0,0 +1,80 @@
+
+        /**
+         * Send error
+         * This method delegate error handling at the server level. This is different
```
delegates

DelazJ commented on python/server/qgsserverresponse.sip:
```diff
@@ -0,0 +1,80 @@
+        /**
+         * Send error
+         * This method delegate error handling at the server level. This is different
+         * from calling setReturnCodei() along with and a specific response body.
```
setReturnCode**i**?

DelazJ commented on python/server/qgsserverresponse.sip:
```diff
@@ -0,0 +1,80 @@
+
+       /**
+        * Write chunk of data
+        * They is a convenient method that will write directly to the 
```
This

DelazJ commented on python/server/qgsserverresponse.sip:
```diff
@@ -0,0 +1,80 @@
+        * Write chunk of data
+        * They is a convenient method that will write directly to the 
+        * underlying I/O device
+        * @return the number of bytes that were actually writtene
```
written

DelazJ commented on python/server/qgsserviceregistry.sip:
```diff
@@ -0,0 +1,83 @@
+ * Class defining the cegistry manager for QGIS server services
+ *
+ * This class provides methods for registering and retrieving 
+ * services. Note that the regstiry does not hord ownership of
```
registry
hold

DelazJ commented on src/server/qgsserverrequest.h:
```diff
@@ -0,0 +1,83 @@
+ */
+
+// Note about design: this intreface must be passed along to python and thus signatures methods must be 
+// compatible with pyQGS/pyQT api and rules.
```
PyQGS or PyQGIS?

DelazJ commented on src/server/qgsserverrequest.h:
```diff
@@ -0,0 +1,83 @@
+       /**
+        * Constructor
+        *
+        * @param url the lurl string
```
lurl?

DelazJ commented on src/server/qgsserverresponse.h:
```diff
@@ -0,0 +1,109 @@
+
+       /** Set Header entry 
+        *  Add Header entry to the response 
+        *  Note that it is usually an error to set Header after writng data
```
writing

DelazJ commented on src/server/qgsserverresponse.h:
```diff
@@ -0,0 +1,109 @@
+        /**
+         * Send error
+         * This method delegate error handling at the server level. This is different
+         * from calling setReturnCodei() along with and a specific response body.
```
setReturnCode**i**?

DelazJ commented on src/server/qgsserverresponse.h:
```diff
@@ -0,0 +1,109 @@
+        virtual void write( const QString& data );
+
+       /**
+        * Write chunk af data
```
of

DelazJ commented on src/server/qgsserverresponse.h:
```diff
@@ -0,0 +1,109 @@
+        * to the underlying I/O device
+        * @return the number of bytes written
+        *
+        * @note not available in pything bindings
```
python?

DelazJ commented on src/server/qgsserviceregistry.h:
```diff
@@ -0,0 +1,108 @@
+/**
+ * \ingroup server
+ * QgsServiceRegistry
+ * Class defining the cegistry manager for QGIS server services
```
registry

DelazJ commented:
Added some spelling-related comments

dmarteau commented on python/server/qgsservice.sip:
```diff
@@ -0,0 +1,60 @@
+     * @param request a QgsServerRequest instance
+     * @param response a QgsServerResponse instance
+     */
+    virtual void executeRequest( const QgsServerRequest& request, QgsServerResponse& response ) = 0;
```
@mhugo yes, I'm not still sure where and how to do that:
- Gives accces from the QgsServerRequest to something like a ServerInstance/Project ?
- Pass a server instance at module registration and access to to the project from it ?
@elpaso ?

mhugo replied:
I'd start with a QgsProject passed here.
But maybe we can refine it when we will start to get rid of the server project parsers and write/adapt real services

dmarteau replied:
@mhugo It is clear thah we will need more than a QgsProject: I suggest to pass a interface from wich you can access the project as suggested by @elpaso.

elpaso replied:
Actually I was thinking at a `QgsServerProjectRegistry` QMap/Hash to hold the `QgsProject` instances.
I also think we'll need a `QgsServerProject` subclass, but I've non checked the details of the new `QgsProject` implementation, maybe we don't need the subclass.

mhugo replied:
The new QgsProject holds all the layers of a project. Do we need more ?

dmarteau replied:
That is the question...

dmarteau replied:
In the current state of the server: is it sufficient to pass a QgsProject if we want to migrate the actual WMS implementation in a registered service ?

mhugo replied:
My current understanding is that a QgsProject will be enough yes, but we (probably @pblottiere ) will tackle this soon normally.

dmarteau replied:
@mhugo Added QgsProject has extra parameter to executeRequest: for now, this seems to be  the simplest and the cleanest way to pass extra parameters to services.

dmarteau committed:
Server refactoring: Added sample native service module; fixed typo

elpaso reviewed (commented):
@dmarteau the unRegisterService is not needed in order to change the configuration at runtime but to change the configuration when the server starts, before it starts serving requests. From a python plugin you could for example unRegister a previously loaded C++ service and use your own implementation instead of the default one.

dmarteau commented on src/server/qgsservicepythonloader.h:
```diff
@@ -0,0 +1,73 @@
+};
+
+#endif
+
```
It is a fact that we can use the plugin mecanism for loading new services: the python tests show that we can effectively load python services from any python code. 
The spirit of this was to not distinguish python modules from natives modules and separate them from the server plugin architecture which target different functional needs. 
I see the idea of 'overriding' default services a more a side effect to the fact that you can access the service API from python plugins. Furtehmore it introduces a difference of status between python and native services: something I would like to avoid.

dmarteau replied:
Another thought: in the context of a server I would consider that the services you want to activate are predefined by some configuration mechanism.

elpaso replied:
This could be hard to implement for **not-core** services if we store this configuration in the project from the OWS option tab: how can QGIS Desktop running on a different machine know what Python services are available?
That's why I believe that core services (WMS, WFS, WCS) should be implemented in C++ and shipped with the QGIS Server package,  and leave Python services for plugins, user-overrides and custom services.

dmarteau replied:
@elpaso From my server administrator/architect point of view I find very strange  for me that the configuration/administration of the server is defined in the project file: it is clearly a security issue.

dmarteau replied:
Yes, perfectly.

dmarteau committed:
fix indent

nyalldawson commented on src/server/qgsserviceregistry.cpp:
```diff
@@ -0,0 +1,166 @@
+// If both segments can be intepreted as numbers the are compared as numbers, otherwise
+// They are compared lexicographically.
+// Return true if v1 is greater than v2
+bool isVersionGreater( const QString& v1, const QString& v2 )
```
I've been thinking about this method - I think this could be moved to a static member in QgsProjectVersion, and then QgsProjectVersion renamed to something generic like QgsVersionNumber. It's quite a handy method which has uses outside of server too.

dmarteau replied:
@nyalldawson that should be another PR ?

nyalldawson replied:
Yeah, not critical and there's a lot of other more important decisions to be made here.

dmarteau commented:
Copied this comment here for better visibility

@elpaso  It is a fact that we can use the plugin mecanism for loading new services: the python tests show that we can effectively load python services from any python code. 
The spirit of this was to not distinguish python modules from natives modules and separate them from the server plugin architecture which target different functional needs. 
I see the idea of 'overriding' default services a more a side effect to the fact that you can access the service API from python plugins. Furthemore it introduces a difference of status between python and native services: something I would like to avoid.

elpaso commented:
@dmarteau I think that kind of separation of C++ core services from Python services is a good thing: ideally a Python service will be eventually migrated to C++ and shipped as a "core" plugin if it's of general interest. Unless we wanted to separate "core" plugins shipped with the official server  (either C++ or Python) from user-provided Python services.  There is another good point in doing so: in case of server embedding, you should have to go from python to the C++ QgsServer and then to Python again for loading the Python services, I've found that this could be difficult because you have to launch a plugin interpreter from within QGIS which is already embedded into Python: GIL lock will prevent you to do this.
I strongly suggest you to follow this implementation:
```
C++ (FCGI) QgSserver:
- start the QgsServer
  - load C++ services
  - load Python plugins
    - register plugin filters using the provided API
    - register Python services from plugin code using the provided API
```

```
Python embedded QgsServer
- start Python
- start  the QgsServer
  - load C++ services
- register Python filters using the provided API
- register Python services using the provided API
```

See an example of hooking plugins into an ambedded server: https://github.com/qgis/QGIS/blob/master/tests/src/python/qgis_wrapped_server.py#L99

dmarteau commented:
@elpaso: I see your point with the embedded server case.

dmarteau committed:
Removed python loader: added unRegisterService() method

m-kuhn commented on python/server/qgsserviceregistry.sip:
```diff
@@ -63,14 +63,22 @@ class QgsServiceRegistry
+     * If the version is not specified then all versions from the specified service 
+     * are unloaded
+     */
+    int unRegisterService( const QString& name, const QString& version = QString() );
```
I think in general this is spelled

     unregisterService

m-kuhn commented on src/server/qgsserverrequest.cpp:
```diff
@@ -40,7 +40,7 @@ QgsServerRequest::~QgsServerRequest()
 }
 
-QString QgsServerRequest::getHeader( const QString& name ) const
+QString QgsServerRequest::getHeader( const QString& /*name*/ ) const
```
Q_UNUSED( name )

dmarteau replied:
This leads to the following compile error:

```C++
/Users/david/Projets/qgis/QGIS/src/server/qgsserverrequest.cpp:43:63: error: expected ')'
QString QgsServerRequest::getHeader( const QString& Q_UNUSED( name ) ) const
                                                              ^
/opt/local/libexec/qt5/include/QtCore/qglobal.h:670:27: note: expanded from macro 'Q_UNUSED'
#define Q_UNUSED(x) (void)x;
             ^
/Users/david/Projets/qgis/QGIS/src/server/qgsserverrequest.cpp:43:36: note: to match this '('
QString QgsServerRequest::getHeader( const QString& Q_UNUSED( name ) ) const
                                   ^
/Users/david/Projets/qgis/QGIS/src/server/qgsserverrequest.cpp:43:27: error: out-of-line declaration of
      'getHeader' does not match any declaration in 'QgsServerRequest'
QString QgsServerRequest::getHeader( const QString& Q_UNUSED( name ) ) const
                          ^~~~~~~~~
/Users/david/Projets/qgis/QGIS/src/server/qgsserverrequest.cpp:43:70: error: expected unqualified-id
QString QgsServerRequest::getHeader( const QString& Q_UNUSED( name ) ) const
```

dmarteau replied:
Arf, misuse of Q_UNUSED :-/

m-kuhn replied:
;)

m-kuhn commented on src/server/qgsserverrequest.cpp:
```diff
@@ -40,7 +40,7 @@ QgsServerRequest::~QgsServerRequest()
-QString QgsServerRequest::getHeader( const QString& name ) const
+QString QgsServerRequest::getHeader( const QString& /*name*/ ) const
 {
   return "";
```
Not sure about this place, but in 90% I use an invalid `QString()` instead of an empty `""`, but really depends on the usecase.

dmarteau committed:
fix typo, use Q_UNUSED, return null/empty QString() instead literal empty string

dmarteau committed:
Add extra parameter QgsProject* to QgsService::executeRequest

dmarteau committed:
Add parameter map accessor from QgsServerRequest

dmarteau committed:
Expose QgsServerInterface to service modules

    * Instanciate QgsServerInterface even if HAVE_SERVER_PYTHON_PLUGINS is
      not defined
    * Pass QgsServerInterface in module registration methods.
    * Add QgsServer initialisation test

 #Please enter the commit message for your changes. Lines starting

sbrunner commented on CMakeLists.txt:
```diff
@@ -445,6 +445,8 @@ IF (WIN32)
   SET (DEFAULT_PLUGIN_SUBDIR  plugins)
   SET (DEFAULT_INCLUDE_SUBDIR include)
 
+  SET (DEFAULT_SERVER_MODULE_SUBDIR server)
```
Should we relay needs a deferent path than the others plugins?

dmarteau replied:
Plugins path is strongly related to plugins for qgis core (loaders....),  I would prefer not to interfere with that atm. In fact, it is actually a quick and easy way to build module's path from QgsApplication::libexecPath().  
Anyway, nothing is fixed now and we can change that later.

sbrunner replied:
ok :-)

sbrunner commented on python/server/qgsserverrequest.sip:
```diff
@@ -0,0 +1,85 @@
+
+    /**
+     *      * @return query params
+     *           */
```
Wrong comments...

sbrunner commented on python/server/qgsserverrequest.sip:
```diff
@@ -0,0 +1,85 @@
+    /**
+     *      * @return query params
+     *           */
+    QMap<QString, QString> parameters() const;
```
Don't we prefer a method like `virtual QString getParameter( const QString& name ) const;`?

dmarteau replied:
I have considered this, but the actual  code seems to favor the usage of a dictionary for parameters so I sticked to that. Another advantage of having, at least internally, a dictionary, it is that you always have to convert your parameter keys and values to uppercase/lowercase You can do that once on the first access so that you can use the same literal strings for keys and values uniformly accross the code. 

And from a python point of view it could be handy to have a dictionary but I do not see any problem to reconsider this choice.

sbrunner replied:
> you always have to convert your parameter keys and values

I don't agree with this sentence, we usually do that in the OGC service because not every body respect exactly the spec, than we do that to increase the compatibility. Than I think that we should be able to write a service that don't do that.

For the initial question I understand your position ant it's true that's will be easier :-)

sbrunner commented on python/server/qgsserverresponse.sip:
```diff
@@ -0,0 +1,80 @@
+     * Send error
+     * This method delegates error handling at the server level. This is different
+     * from calling setReturnCode() along with and a specific response body.
+     * @param code HHTP return code value
```
=> HTTP

sbrunner commented on python/server/qgsserverresponse.sip:
```diff
@@ -0,0 +1,80 @@
+     * @param code HHTP return code value
+     * @param message An informative error message
+     */
+    virtual void sendError( int code,  const QString& message ) = 0;
```
it's not a set?

sbrunner replied:
Retarding the other comment it definitively not a set :-)

sbrunner commented on python/server/qgsserverresponse.sip:
```diff
@@ -0,0 +1,80 @@
+    /** Set the http return code
+     * @param code HTTP return code value
+     */
+    virtual void setReturnCode( int code ) = 0;
```
shouldn't be setErrorCode to be coherent with the next method?

dmarteau replied:
HTTP return code is not necessary an error code: you can return code <> 200 with legitimate content. This is different that sendError( ) which delegate the error handling to the transport/IO framework. sendError() should end the transaction and any attempt to write data or set headers should be an error.  Maybe the comments should be completed.

sbrunner replied:
OK, thanks for your explain :-) I also read a bit too fast... Then to be complete shouldn't we also be able to change the message (relay not sure that needed)?

sbrunner commented on src/server/qgsserver.h:
```diff
@@ -59,7 +60,7 @@ class SERVER_EXPORT QgsServer
      * @param val value
      * @note added in 2.14
-     */
+    / */
```
to be removed ...

sbrunner reviewed (commented):
Nice work :-)

dmarteau commented:
@sbrunner  thanks :-)

dmarteau committed:
Fix coding style and update comments

dmarteau committed:
Server: Brut force port WMS as service module

    - Compile and load as service but functionality not tested
    - Removed dependencies to QgsRequestHandler
    - Removed dependencies to QgsOWSServer
    - Separate requests in their own units

dmarteau committed:
Fix coding style

dmarteau committed:
Fix 'Incomplete type' compile error

dmarteau committed:
fix prefix path setting: enable running from from build directory

dmarteau commented:
Actually the test pass on osx so I have no clue why it fails on travis ( and why it is followed with a segmentation fault) .

dmarteau committed:
Fix spelling

dmarteau committed:
Create module test; add more debug messages for travis

dmarteau committed:
Fix indent

dmarteau committed:
Fix DEFAULT_SERVER_MODULE_SUBDIR in CMakeLists.txt

rldhont reviewed (approved):
Thanks for porting WMS

sbrunner reviewed (approved):
Create Job :-)

dmarteau commented:
Things left to do:

1. Wrap actual request handler with a QgsServerResponse adapter
1. Call the service module on WMS requests.
1. Remove WMS code from server main module.

rldhont commented:
@pblottiere it will be interesting to synchronize works. https://github.com/qgis/QGIS/pull/3942

dmarteau commented:
@pblottiere actually the WMS code in the module is 98% identical to the original WMS code so porting your code should not be a problem. Thing to take care is that RequestHandler is no longer used and some server components may required to be exported. 
I would suggest to wait for WMS beeing a fully functional service before porting your changes: I want to be sure that we reach iso-functionality with the service module before dealing with more changes.

pblottiere commented:
@dmarteau 

> actually the WMS code in the module is 98% identical to the original WMS code so porting your code should not be a problem. 

Yes, I agree pretty much with you.

> I would suggest to wait for WMS beeing a fully functional service before porting your changes: I want to be sure that we reach iso-functionality with the service module before dealing with more changes.

OK. Anyway, I have some work to do to completely remove project parsing from the server.

dmarteau committed:
Implement QgsFcgiRequest and QgsFcgiResponse

dmarteau commented:
@elpaso @rldhont I have introduced some changes to the QgsRequestHandler interface: can you tell if these are ok for you ?

elpaso commented on src/server/qgsrequesthandler.h:
```diff
@@ -96,33 +96,26 @@ class QgsRequestHandler
-    //! Send out HTTP headers and flush output buffer
+    /** Send out HTTP headers and flush output buffer
+     *
+     *  This method is intended osly for streaming
```
typo: only

dmarteau commented on src/server/qgsrequesthandler.h:
```diff
@@ -96,33 +96,26 @@ class QgsRequestHandler
     virtual void appendBody( const QByteArray &body ) = 0;
 
-    //! Clears the response body
-    virtual void clearBody() = 0;
```
I was expecting this question :-)

Body and headers are related to each other, removing body without clearing the headers or removing all headers and keeping the whole body does not really make sense functionaly: I can't think of a use case where you would wipe out one without wiping the other - (As an example: all tests that where calling clear[Headers|Body] were calling both).

This is clearly related to what plugins can do with the content of the response: either they modify the content of the actual body and thus they don't really have to change the actual headers  or they build a completely new content and thus may eventually need to clear everything.

dmarteau replied:
"BTW you can always get the list of headers keys and clear them one by one from a plugin":
As a matter af fact no :  this possibility did not even existed before (see the QgsRequestHandler interface in master) so there is little advantage to enable it.

In fact yes, there is a drawback: you need to implement it ! 
We may be hitting a philosophical point here :-). My point of view is: keep code minimal, if you don't see the use for a feature, do not implement it - until it is required.

Note that from now, the response parameters are abstracted in QgsServerResponse objects (an abstract QgsRequestHandler is no more needed) and I really would like to keep the its interface minimal.

elpaso replied:
mmm you can remove a header, but you need to know its name, http://qgis.org/api/classQgsRequestHandler.html#a64f07891172cd1c73d760947c531f418
I'd probably like to see a method to return the header keys as a list so we can clear a single one in case we need to.

About the minimalistic approach, I tend to apply it to the code that I'm supposed to not distribute and over which I have full control of the release cycle, but in this case you need to consider that we are designing an API over an existing one, and that we cannot foresee all the possible use cases and I'm pretty sure that if we cut some of the existing functionality somebody will promptly find a use case for it. 
This of course does not mean we should implement all possible corner cases, but something in the middle would probably fit best.

On public APIs I tend to appreciate a lot rich APIs whre the wise designers leave to the client code the possibility to alter whatever you want without breaking the internals of the class.

Back to the point: it's fine for me to drop the clear* methods if there is a way to do that by clearing the individual items in the header, which is desirable in any case, so I'd add a method to retrieve the header "keys".

dmarteau replied:
Removing/adding individual header is still here, so, fundamentally, nothing has changed  from this perspective. 

Instead a 'keys' method I would favor returning a Map holding key/values it would be more practical: the use case for this could be introspection/debugging from plugins.

elpaso replied:
As long as you are returning a not-const pointer that can alter the headers `QMap` and if it's available in the bindings, it looks good to me,  a SIP mapped type for `QMap<QString, QString>` should be already available.

I can't recall why I did not write such an accessor to the `QMap` headers when I first designed this API, maybe I did not want to leave direct manipulation of the headers `QMap` dictionary from client code.

If you write such an accessor, there will be no need for any header manipulation methods (setHeader, removeHeader, clearHeaders) since the client code could directly alter the headers dictionary.

Just to be sure we are on the same page: as long as the possibility to create/edit/delete/retrieve individual header entries from client code and plugins is maintained, choose whatever implementation you feel suit best.

elpaso replied:
So, if there is a good reason to not let the client code access the headers directly in write mode, it's mandatory to keep the getter/setters and add the missing keys() method. Why?
Because passing a read-only (or copy) object to python bindings is not what the developer expects: if I get a dictionary object in Python I expect I can alter it. Passing a tuple of tuples or some frozen struct doesn't seem the right thing to do either. 
Remember: I want to be able to create/edit/delete/retrieve individual header entries from client code and plugins.

dmarteau replied:
@elpaso "Just to be sure we are on the same page: as long as the possibility to create/edit/delete/retrieve individual header entries from client code and plugins is maintained":  for information, the original implementation has only setHeader/removeHeader methods (besside clearHeaders).

elpaso replied:
that made it possible to create/edit/delete individual headers, the current implementation only misses the ability to get the list of header names and retrieve a single one, which should be added now.

dmarteau replied:
done.

dmarteau committed:
QgsRequestHandler is now a concrete class

    I/O transactions are now abstracted with QgsServerRequest/QgsServerResponse
    So there is no need for another abstraction layer

dmarteau committed:
Add headers keys and value accessors

dmarteau committed:
Ensure url is complete in fcgi request

dmarteau committed:
Add plugin hook support decorator

    - Add plugin hook support decorator for use in  service modules
    - Remove handling of plugin hook with request handler (not needed
anymore)
    - Reduce usage of HAVE_SERVER_PYTHON_PLUGINS by using forward opaque declaration.

dmarteau committed:
Port wmsserver changes to service

dmarteau committed:
fix doc warning

dmarteau committed:
Fix indent and documentation

dmarteau committed:
Fix sip syntax error

dmarteau committed:
Port bugfix from PR #3906 to service

dmarteau committed:
WMS run as a module service !

elpaso commented on src/server/qgsbufferserverresponse.h:
```diff
@@ -42,6 +42,12 @@ class QgsBufferServerResponse: public QgsServerResponse
+
+    virtual QList<QString> headerKeys() const override;
+
+    virtual bool headersWritten() const override;
```
May I suggest `headersSent` instead of written? I think it's more intuitive and common among other frameworks and languages.

dmarteau replied:
No problem

rldhont commented:
@dmarteau thanks for 84400fe!
I have to prot the other WxS as service

dmarteau commented:
@elpaso So far, the wms is now migrated to its own module and it is operational. There is still some cleanup to do:
- Remove code specific to wms from the the server code unit.
- Replace all calls to getenv and use parameter's accessor instead (using getenv for accessing parameters is very fcgi centric).
 - Move others services to their own modules.

Each of these can be done independently without affecting functionality and I would suggest that these modifications go along with their specific PR.

Furthemore, in the actual state it is going to be painful to stay in sync with every changes made to the WMS  and maintain to version of the code in this branch.  I everybody agree, I would consider that this PR has reached is goal and would ask for a final review.

rldhont commented:
+1 for merging

m-kuhn reviewed (dismissed):
Please remove the Testing file before merging.

dmarteau committed:
Remove Testing/Temporary/CTestCostData.txt

dmarteau committed:
Replace headersWritten by headersSent

dmarteau committed:
Remove wmsserver files from server code unit

rldhont commented:
@m-kuhn Testing files has been removed. Is it ok for you to merge now ?
@elpaso are you also ok for merging now ?

elpaso commented:
@rldhont @dmarteau  +1 for merge: this is an amazing job, thanks!
We can fix the tiny bits later if we need to.
For the environment passing, it's fne for me to move it in the request object, but please keep the possibility to get QUERY_STRING and other vars from the env, this is also very useful when running the server from the command line for testing.

dmarteau commented:
@elpaso thanks. About getenv: was thinking as things like https://github.com/dmarteau/QGIS/blob/master/src/server/services/wms/qgswmsgetcapabilities.cpp#L39 or https://github.com/dmarteau/QGIS/blob/master/src/server/services/wms/qgswmsservertransitional.cpp#L2666 where informations should be retrieved from the request url object, not getenv.

elpaso commented:
Sorry @dmarteau  maybe I don't understand completely here.
If you plan to move the env vars outside from the server modules code into the request object (not the request URL!) then I completely agree, all inputs should go into the request object and environment is certainly input. 

If not, how do you plan to pass the server name SERVER_NAME from the web server tier if not using the environment? AFAIK these informations (like the protocol, port, if is secure etc.) will always come from the web server (Apache, nginx, whatever), how do you plan to pass them into the services?

I think that the request object should take care of collecting the SERVER_NAME and all other variables from the environment, but we can imagine that a specialized sub-class of request object could collect the same information from other sources though.

So, big +1 to move all the input related functions out of the services and into the request classes.

dmarteau commented:
@elpaso 'So, big +1 to move all the input related functions out of the services and into the request classes': yes this is exactly what I have in mind :-)

dmarteau commented:
@elpaso @rldhont  do net merge before #3886 !

dmarteau commented:
Please somebody can fix VectorTest ??

dmarteau commented:
@volaya  please can you fix https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tools/vector.py ?

dmarteau commented:
Processing is still broken....

m-kuhn commented:
> @m-kuhn Testing files has been removed. Is it ok for you to merge now ?

It's the only thing I have really looked at. While I am very (!) excited about this happening, I think there have been enough eyeballs on this that it doesn't also need my review. So if everyone else is happy, please proceed and merge.

nyalldawson commented on python/server/qgsrequesthandler.sip:
```diff
@@ -61,57 +61,52 @@ class QgsRequestHandler
     // virtual void setGetFeatureResponse( QByteArray* ba ) = 0;
 
     //! @note not available in Python bindings
```
Why are these marked not available in python?

dmarteau replied:
@nyalldawson I don't know, this has been left unchanged. My felling it is that they are ad-hoc methods used only in wfs for streaming getfeature request. IMHO they brings very little functionality as they just encapsulate write/flush to response and I think also that they are misplaced.

I have planned a PR to make some cleanup when this one will be merged.

nyalldawson commented on src/server/qgsbufferserverresponse.cpp:
```diff
@@ -0,0 +1,161 @@
+QgsBufferServerResponse::QgsBufferServerResponse()
+{
+  mBuffer.open( QIODevice::ReadWrite );
+  mHeadersSent = false;
```
Should either be an initializer list, or even better c++11  class member initializers (Define value in header - see http://www.informit.com/articles/article.aspx?p=1852519)

nyalldawson commented on src/server/qgsbufferserverresponse.h:
```diff
@@ -0,0 +1,125 @@
+
+/**
+ * \ingroup server
+ * QgsBufferServerResponse
```
This should be

```
* \ingroup server
* \class QgsBufferServerResponse
* Class defining buffered response
*/

Otherwise the doxygen brief comment for the class will just be "QgsBufferServerResponse", rather then the descriptive text.

nyalldawson commented on src/server/qgsfilterresponsedecorator.h:
```diff
@@ -0,0 +1,78 @@
+    ~QgsFilterResponseDecorator();
+
+    /**
+     * Call flter's requestReady() method
```
filters

nyalldawson commented on src/server/qgsrequesthandler.cpp:
```diff
@@ -643,75 +535,80 @@ void QgsHttpRequestHandler::requestStringToParameterMap( const QString& request,
+
+    QDomDocument doc;
+    QString errorMsg;
+    int line;
```
should be initialized for safety

nyalldawson commented on src/server/qgsserver.cpp:
```diff
@@ -493,81 +458,122 @@ QPair<QByteArray, QByteArray> QgsServer::handleRequest( const QString& queryStri
         wfsServer.executeRequest();
       }
     }
+    /*
```
Please #ifdef 0 this out rather than using comments

nyalldawson commented on src/server/qgsserverrequest.h:
```diff
@@ -0,0 +1,135 @@
+ * Class defining request interface passed to services QgsService::executeRequest() method
+ */
+
+// Note about design: this intreface must be passed along to python and thus signatures methods must be
```
interface

nyalldawson commented on src/server/qgsserverresponse.h:
```diff
@@ -0,0 +1,153 @@
+
+    /**
+     * Clear header
+     * Undo a previous 'set_header' call
```
setHeader

nyalldawson commented on src/server/qgsservice.h:
```diff
@@ -0,0 +1,73 @@
+    QgsService();
+
+    //! Destructor
+    virtual ~QgsService();
```
Use 

    virtual ~QgsService() = default;

instead of an empty implementation

nyalldawson commented:
@dmarteau just some small code quality changes requested

dmarteau committed:
Code quality changes and typo

rldhont commented:
Merge done

nyalldawson commented:
Brilliant work David!

dmarteau commented:
@nyalldawson thanks.

sbrunner commented:
Many thanks for your great job :-)

elpaso commented:
Great job @dmarteau, thanks a lot!

m-kuhn commented:
Yay, good job :+1: ! 3.0 will be THE release!

elpaso commented on src/server/qgsserver.cpp:
```diff
@@ -394,97 +363,93 @@ QPair<QByteArray, QByteArray> QgsServer::handleRequest( const QString& queryStri
   QString configFilePath = configPath( *sConfigFilePath, parameterMap );
+
 #ifdef HAVE_SERVER_PYTHON_PLUGINS
+  // XXX Why this is enabled only fol plugins ?
```
because when Python plugins are disabled there is no sServerInterface
